File: AIAgent/ZBuild/Repos/cxdb-main/clients/go/types/builders.go
// Copyright 2025 StrongDM Inc
// SPDX-License-Identifier: Apache-2.0

package types

// =============================================================================
// Context Metadata Helpers
// =============================================================================

// WithContextMetadata attaches context metadata to any ConversationItem.
// By convention, this should only be used on the first turn of a context.
func (item *ConversationItem) WithContextMetadata(meta *ContextMetadata) *ConversationItem {
	item.ContextMetadata = meta
	return item
}

// WithClientTag is a convenience method to attach just a client tag.
// By convention, this should only be used on the first turn of a context.
func (item *ConversationItem) WithClientTag(tag string) *ConversationItem {
	if item.ContextMetadata == nil {
		item.ContextMetadata = &ContextMetadata{}
	}
	item.ContextMetadata.ClientTag = tag
	return item
}

// =============================================================================
// User Input Builders
// =============================================================================

// NewUserInput creates a user input conversation item.
func NewUserInput(text string, files ...string) *ConversationItem {
	return &ConversationItem{
		ItemType:  ItemTypeUserInput,
		Status:    ItemStatusComplete,
		Timestamp: Now(),
		UserInput: &UserInput{
			Text:  text,
			Files: files,
		},
	}
}

// =============================================================================
// Assistant Turn Builders (v2 - preferred)
// =============================================================================

// NewAssistantTurn creates an assistant turn conversation item.
func NewAssistantTurn(text string) *ConversationItem {
	return &ConversationItem{
		ItemType:  ItemTypeAssistantTurn,
		Status:    ItemStatusComplete,
		Timestamp: Now(),
		Turn: &AssistantTurn{
			Text: text,
		},
	}
}

// AssistantTurnBuilder provides fluent configuration for assistant turn items.
type AssistantTurnBuilder struct {
	item *ConversationItem
}

// BuildAssistantTurn starts building an assistant turn conversation item.
func BuildAssistantTurn(text string) *AssistantTurnBuilder {
	return &AssistantTurnBuilder{
		item: &ConversationItem{
			ItemType:  ItemTypeAssistantTurn,
			Status:    ItemStatusComplete,
			Timestamp: Now(),
			Turn: &AssistantTurn{
				Text:      text,
				ToolCalls: make([]ToolCallItem, 0),
			},
		},
	}
}

// WithReasoning adds reasoning/thinking content.
func (b *AssistantTurnBuilder) WithReasoning(reasoning string) *AssistantTurnBuilder {
	b.item.Turn.Reasoning = reasoning
	return b
}

// WithAgent sets the agent name.
func (b *AssistantTurnBuilder) WithAgent(agent string) *AssistantTurnBuilder {
	b.item.Turn.Agent = agent
	return b
}

// WithTurnNumber sets the turn number and optionally max turns.
func (b *AssistantTurnBuilder) WithTurnNumber(turn, maxTurns int) *AssistantTurnBuilder {
	b.item.Turn.TurnNumber = turn
	b.item.Turn.MaxTurns = maxTurns
	return b
}

// WithFinishReason sets the finish reason.
func (b *AssistantTurnBuilder) WithFinishReason(reason string) *AssistantTurnBuilder {
	b.item.Turn.FinishReason = reason
	return b
}

// WithMetrics sets token usage metrics.
func (b *AssistantTurnBuilder) WithMetrics(inputTokens, outputTokens int64) *AssistantTurnBuilder {
	b.item.Turn.Metrics = &TurnMetrics{
		InputTokens:  inputTokens,
		OutputTokens: outputTokens,
		TotalTokens:  inputTokens + outputTokens,
	}
	return b
}

// WithFullMetrics sets complete token usage metrics.
func (b *AssistantTurnBuilder) WithFullMetrics(metrics *TurnMetrics) *AssistantTurnBuilder {
	b.item.Turn.Metrics = metrics
	return b
}

// WithToolCall adds a tool call to the turn.
func (b *AssistantTurnBuilder) WithToolCall(tc ToolCallItem) *AssistantTurnBuilder {
	b.item.Turn.ToolCalls = append(b.item.Turn.ToolCalls, tc)
	return b
}

// WithStatus sets the item status.
func (b *AssistantTurnBuilder) WithStatus(status ItemStatus) *AssistantTurnBuilder {
	b.item.Status = status
	return b
}

// WithID sets the item ID.
func (b *AssistantTurnBuilder) WithID(id string) *AssistantTurnBuilder {
	b.item.ID = id
	return b
}

// WithContextMetadata attaches context metadata (for first turn only).
func (b *AssistantTurnBuilder) WithContextMetadata(meta *ContextMetadata) *AssistantTurnBuilder {
	b.item.ContextMetadata = meta
	return b
}

// WithClientTag attaches a client tag (for first turn only).
func (b *AssistantTurnBuilder) WithClientTag(tag string) *AssistantTurnBuilder {
	if b.item.ContextMetadata == nil {
		b.item.ContextMetadata = &ContextMetadata{}
	}
	b.item.ContextMetadata.ClientTag = tag
	return b
}

// Build returns the configured conversation item.
func (b *AssistantTurnBuilder) Build() *ConversationItem {
	return b.item
}

// =============================================================================
// Tool Call Item Builder
// =============================================================================

// NewToolCallItem creates a tool call item for embedding in an assistant turn.
func NewToolCallItem(id, name, args string) ToolCallItem {
	return ToolCallItem{
		ID:     id,
		Name:   name,
		Args:   args,
		Status: ToolCallStatusPending,
	}
}

// ToolCallItemBuilder provides fluent configuration for tool call items.
type ToolCallItemBuilder struct {
	tc ToolCallItem
}

// BuildToolCallItem starts building a tool call item.
func BuildToolCallItem(id, name, args string) *ToolCallItemBuilder {
	return &ToolCallItemBuilder{
		tc: ToolCallItem{
			ID:     id,
			Name:   name,
			Args:   args,
			Status: ToolCallStatusPending,
		},
	}
}

// WithDescription adds a human-readable description.
func (b *ToolCallItemBuilder) WithDescription(desc string) *ToolCallItemBuilder {
	b.tc.Description = desc
	return b
}

// WithStatus sets the tool call status.
func (b *ToolCallItemBuilder) WithStatus(status ToolCallStatus) *ToolCallItemBuilder {
	b.tc.Status = status
	return b
}

// WithStreamingOutput sets accumulated streaming output.
func (b *ToolCallItemBuilder) WithStreamingOutput(output string, truncated bool) *ToolCallItemBuilder {
	b.tc.StreamingOutput = output
	b.tc.StreamingOutputTruncated = truncated
	return b
}

// WithResult sets the successful result.
func (b *ToolCallItemBuilder) WithResult(content string, exitCode *int) *ToolCallItemBuilder {
	b.tc.Status = ToolCallStatusComplete
	b.tc.Result = &ToolCallResult{
		Content:  content,
		Success:  true,
		ExitCode: exitCode,
	}
	return b
}

// WithError sets the error result.
func (b *ToolCallItemBuilder) WithError(message string, exitCode *int) *ToolCallItemBuilder {
	b.tc.Status = ToolCallStatusError
	b.tc.Error = &ToolCallError{
		Message:  message,
		ExitCode: exitCode,
	}
	return b
}

// WithDuration sets the execution duration in milliseconds.
func (b *ToolCallItemBuilder) WithDuration(ms int64) *ToolCallItemBuilder {
	b.tc.DurationMs = ms
	return b
}

// Build returns the configured tool call item.
func (b *ToolCallItemBuilder) Build() ToolCallItem {
	return b.tc
}

// =============================================================================
// Handoff Builder
// =============================================================================

// NewHandoff creates a handoff conversation item.
func NewHandoff(fromAgent, toAgent string) *ConversationItem {
	return &ConversationItem{
		ItemType:  ItemTypeHandoff,
		Status:    ItemStatusComplete,
		Timestamp: Now(),
		Handoff: &HandoffInfo{
			FromAgent: fromAgent,
			ToAgent:   toAgent,
		},
	}
}

// HandoffBuilder provides fluent configuration for handoff items.
type HandoffBuilder struct {
	item *ConversationItem
}

// BuildHandoff starts building a handoff conversation item.
func BuildHandoff(fromAgent, toAgent string) *HandoffBuilder {
	return &HandoffBuilder{
		item: &ConversationItem{
			ItemType:  ItemTypeHandoff,
			Status:    ItemStatusComplete,
			Timestamp: Now(),
			Handoff: &HandoffInfo{
				FromAgent: fromAgent,
				ToAgent:   toAgent,
			},
		},
	}
}

// WithToolName sets the handoff tool name.
func (b *HandoffBuilder) WithToolName(toolName string) *HandoffBuilder {
	b.item.Handoff.ToolName = toolName
	return b
}

// WithInput sets the input passed to the target agent.
func (b *HandoffBuilder) WithInput(input string) *HandoffBuilder {
	b.item.Handoff.Input = input
	return b
}

// WithReason sets the reason for the handoff.
func (b *HandoffBuilder) WithReason(reason string) *HandoffBuilder {
	b.item.Handoff.Reason = reason
	return b
}

// WithID sets the item ID.
func (b *HandoffBuilder) WithID(id string) *HandoffBuilder {
	b.item.ID = id
	return b
}

// Build returns the configured conversation item.
func (b *HandoffBuilder) Build() *ConversationItem {
	return b.item
}

// =============================================================================
// System Message Builders
// =============================================================================

// NewSystemInfo creates an informational system message.
func NewSystemInfo(content string) *ConversationItem {
	return &ConversationItem{
		ItemType:  ItemTypeSystem,
		Status:    ItemStatusComplete,
		Timestamp: Now(),
		System: &SystemMessage{
			Kind:    SystemKindInfo,
			Content: content,
		},
	}
}

// NewSystemWarning creates a warning system message.
func NewSystemWarning(content string) *ConversationItem {
	return &ConversationItem{
		ItemType:  ItemTypeSystem,
		Status:    ItemStatusComplete,
		Timestamp: Now(),
		System: &SystemMessage{
			Kind:    SystemKindWarning,
			Content: content,
		},
	}
}

// NewSystemError creates an error system message.
func NewSystemError(content string) *ConversationItem {
	return &ConversationItem{
		ItemType:  ItemTypeSystem,
		Status:    ItemStatusComplete,
		Timestamp: Now(),
		System: &SystemMessage{
			Kind:    SystemKindError,
			Content: content,
		},
	}
}

// SystemBuilder provides fluent configuration for system message items.
type SystemBuilder struct {
	item *ConversationItem
}

// BuildSystem starts building a system message conversation item.
func BuildSystem(kind SystemKind, content string) *SystemBuilder {
	return &SystemBuilder{
		item: &ConversationItem{
			ItemType:  ItemTypeSystem,
			Status:    ItemStatusComplete,
			Timestamp: Now(),
			System: &SystemMessage{
				Kind:    kind,
				Content: content,
			},
		},
	}
}

// WithTitle adds a title to the system message.
func (b *SystemBuilder) WithTitle(title string) *SystemBuilder {
	b.item.System.Title = title
	return b
}

// WithID sets the item ID.
func (b *SystemBuilder) WithID(id string) *SystemBuilder {
	b.item.ID = id
	return b
}

// Build returns the configured conversation item.
func (b *SystemBuilder) Build() *ConversationItem {
	return b.item
}

// =============================================================================
// Legacy Builders (v1 schema - kept for backward compatibility)
// =============================================================================

// NewAssistant creates an assistant response conversation item (legacy).
// Deprecated: Use NewAssistantTurn for new code.
func NewAssistant(text string) *ConversationItem {
	return &ConversationItem{
		ItemType:  ItemTypeAssistant,
		Status:    ItemStatusComplete,
		Timestamp: Now(),
		Assistant: &Assistant{
			Text: text,
		},
	}
}

// AssistantBuilder provides fluent configuration for assistant items (legacy).
// Deprecated: Use AssistantTurnBuilder for new code.
type AssistantBuilder struct {
	item *ConversationItem
}

// BuildAssistant starts building an assistant conversation item (legacy).
// Deprecated: Use BuildAssistantTurn for new code.
func BuildAssistant(text string) *AssistantBuilder {
	return &AssistantBuilder{
		item: &ConversationItem{
			ItemType:  ItemTypeAssistant,
			Status:    ItemStatusComplete,
			Timestamp: Now(),
			Assistant: &Assistant{
				Text: text,
			},
		},
	}
}

// WithReasoning adds reasoning/thinking content.
func (b *AssistantBuilder) WithReasoning(reasoning string) *AssistantBuilder {
	b.item.Assistant.Reasoning = reasoning
	return b
}

// WithModel sets the model name.
func (b *AssistantBuilder) WithModel(model string) *AssistantBuilder {
	b.item.Assistant.Model = model
	return b
}

// WithTokens sets token usage metrics.
func (b *AssistantBuilder) WithTokens(input, output int64) *AssistantBuilder {
	b.item.Assistant.InputTokens = input
	b.item.Assistant.OutputTokens = output
	return b
}

// WithStopReason sets the stop reason.
func (b *AssistantBuilder) WithStopReason(reason string) *AssistantBuilder {
	b.item.Assistant.StopReason = reason
	return b
}

// WithStatus sets the item status.
func (b *AssistantBuilder) WithStatus(status ItemStatus) *AssistantBuilder {
	b.item.Status = status
	return b
}

// Build returns the configured conversation item.
func (b *AssistantBuilder) Build() *ConversationItem {
	return b.item
}

// NewToolCall creates a tool call conversation item (legacy).
// Deprecated: Use BuildAssistantTurn().WithToolCall() for new code.
func NewToolCall(callID, name, args string) *ConversationItem {
	return &ConversationItem{
		ItemType:  ItemTypeToolCall,
		Status:    ItemStatusPending,
		Timestamp: Now(),
		ToolCall: &ToolCall{
			CallID: callID,
			Name:   name,
			Args:   args,
		},
	}
}

// ToolCallBuilder provides fluent configuration for tool call items (legacy).
// Deprecated: Use ToolCallItemBuilder for new code.
type ToolCallBuilder struct {
	item *ConversationItem
}

// BuildToolCall starts building a tool call conversation item (legacy).
// Deprecated: Use BuildToolCallItem for new code.
func BuildToolCall(callID, name, args string) *ToolCallBuilder {
	return &ToolCallBuilder{
		item: &ConversationItem{
			ItemType:  ItemTypeToolCall,
			Status:    ItemStatusPending,
			Timestamp: Now(),
			ToolCall: &ToolCall{
				CallID: callID,
				Name:   name,
				Args:   args,
			},
		},
	}
}

// WithDescription adds a human-readable description.
func (b *ToolCallBuilder) WithDescription(desc string) *ToolCallBuilder {
	b.item.ToolCall.Description = desc
	return b
}

// WithStatus sets the item status.
func (b *ToolCallBuilder) WithStatus(status ItemStatus) *ToolCallBuilder {
	b.item.Status = status
	return b
}

// Build returns the configured conversation item.
func (b *ToolCallBuilder) Build() *ConversationItem {
	return b.item
}

// NewToolResult creates a tool result conversation item (legacy).
// Deprecated: Use BuildAssistantTurn().WithToolCall() with Result for new code.
func NewToolResult(callID, content string, isError bool) *ConversationItem {
	return &ConversationItem{
		ItemType:  ItemTypeToolResult,
		Status:    ItemStatusComplete,
		Timestamp: Now(),
		ToolResult: &ToolResult{
			CallID:  callID,
			Content: content,
			IsError: isError,
		},
	}
}

// ToolResultBuilder provides fluent configuration for tool result items (legacy).
// Deprecated: Use ToolCallItemBuilder with WithResult/WithError for new code.
type ToolResultBuilder struct {
	item *ConversationItem
}

// BuildToolResult starts building a tool result conversation item (legacy).
// Deprecated: Use BuildToolCallItem with WithResult for new code.
func BuildToolResult(callID, content string) *ToolResultBuilder {
	return &ToolResultBuilder{
		item: &ConversationItem{
			ItemType:  ItemTypeToolResult,
			Status:    ItemStatusComplete,
			Timestamp: Now(),
			ToolResult: &ToolResult{
				CallID:  callID,
				Content: content,
				IsError: false,
			},
		},
	}
}

// WithError marks this as an error result.
func (b *ToolResultBuilder) WithError() *ToolResultBuilder {
	b.item.ToolResult.IsError = true
	return b
}

// WithExitCode sets the exit code for shell commands.
func (b *ToolResultBuilder) WithExitCode(code int) *ToolResultBuilder {
	b.item.ToolResult.ExitCode = &code
	return b
}

// WithStreamingOutput sets accumulated streaming output.
func (b *ToolResultBuilder) WithStreamingOutput(output string) *ToolResultBuilder {
	b.item.ToolResult.StreamingOutput = output
	return b
}

// WithTruncated marks the output as truncated.
func (b *ToolResultBuilder) WithTruncated() *ToolResultBuilder {
	b.item.ToolResult.OutputTruncated = true
	return b
}

// WithDuration sets the execution duration in milliseconds.
func (b *ToolResultBuilder) WithDuration(ms int64) *ToolResultBuilder {
	b.item.ToolResult.DurationMs = ms
	return b
}

// Build returns the configured conversation item.
func (b *ToolResultBuilder) Build() *ConversationItem {
	return b.item
}

File: AIAgent/ZBuild/Repos/cxdb-main/clients/rust/src/types/builders.rs
// Copyright 2025 StrongDM Inc
// SPDX-License-Identifier: Apache-2.0

use super::conversation::*;
use super::provenance::Provenance;

impl ConversationItem {
    pub fn with_context_metadata(&mut self, meta: ContextMetadata) -> &mut Self {
        self.context_metadata = Some(meta);
        self
    }

    pub fn with_client_tag(&mut self, tag: impl Into<String>) -> &mut Self {
        if self.context_metadata.is_none() {
            self.context_metadata = Some(ContextMetadata {
                client_tag: String::new(),
                title: String::new(),
                labels: Vec::new(),
                custom: std::collections::HashMap::new(),
                provenance: None,
            });
        }
        if let Some(meta) = &mut self.context_metadata {
            meta.client_tag = tag.into();
        }
        self
    }
}

pub fn new_user_input(text: impl Into<String>, files: Vec<String>) -> ConversationItem {
    ConversationItem {
        item_type: ItemTypeUserInput.to_string(),
        status: ItemStatusComplete.to_string(),
        timestamp: Now(),
        id: String::new(),
        user_input: Some(UserInput {
            text: text.into(),
            files,
        }),
        turn: None,
        system: None,
        handoff: None,
        assistant: None,
        tool_call: None,
        tool_result: None,
        context_metadata: None,
    }
}

pub fn new_assistant_turn(text: impl Into<String>) -> ConversationItem {
    ConversationItem {
        item_type: ItemTypeAssistantTurn.to_string(),
        status: ItemStatusComplete.to_string(),
        timestamp: Now(),
        id: String::new(),
        user_input: None,
        turn: Some(AssistantTurn {
            text: text.into(),
            tool_calls: Vec::new(),
            reasoning: String::new(),
            metrics: None,
            agent: String::new(),
            turn_number: 0,
            max_turns: 0,
            finish_reason: String::new(),
        }),
        system: None,
        handoff: None,
        assistant: None,
        tool_call: None,
        tool_result: None,
        context_metadata: None,
    }
}

pub struct AssistantTurnBuilder {
    item: ConversationItem,
}

pub fn build_assistant_turn(text: impl Into<String>) -> AssistantTurnBuilder {
    AssistantTurnBuilder {
        item: new_assistant_turn(text),
    }
}

impl AssistantTurnBuilder {
    pub fn with_reasoning(&mut self, reasoning: impl Into<String>) -> &mut Self {
        if let Some(turn) = &mut self.item.turn {
            turn.reasoning = reasoning.into();
        }
        self
    }

    pub fn with_agent(&mut self, agent: impl Into<String>) -> &mut Self {
        if let Some(turn) = &mut self.item.turn {
            turn.agent = agent.into();
        }
        self
    }

    pub fn with_turn_number(&mut self, turn: i64, max_turns: i64) -> &mut Self {
        if let Some(turn_item) = &mut self.item.turn {
            turn_item.turn_number = turn;
            turn_item.max_turns = max_turns;
        }
        self
    }

    pub fn with_finish_reason(&mut self, reason: impl Into<String>) -> &mut Self {
        if let Some(turn) = &mut self.item.turn {
            turn.finish_reason = reason.into();
        }
        self
    }

    pub fn with_metrics(&mut self, input_tokens: i64, output_tokens: i64) -> &mut Self {
        if let Some(turn) = &mut self.item.turn {
            turn.metrics = Some(TurnMetrics {
                input_tokens,
                output_tokens,
                total_tokens: input_tokens + output_tokens,
                cached_tokens: None,
                reasoning_tokens: None,
                duration_ms: None,
                model: String::new(),
            });
        }
        self
    }

    pub fn with_full_metrics(&mut self, metrics: TurnMetrics) -> &mut Self {
        if let Some(turn) = &mut self.item.turn {
            turn.metrics = Some(metrics);
        }
        self
    }

    pub fn with_tool_call(&mut self, tool_call: ToolCallItem) -> &mut Self {
        if let Some(turn) = &mut self.item.turn {
            turn.tool_calls.push(tool_call);
        }
        self
    }

    pub fn with_status(&mut self, status: impl Into<String>) -> &mut Self {
        self.item.status = status.into();
        self
    }

    pub fn with_id(&mut self, id: impl Into<String>) -> &mut Self {
        self.item.id = id.into();
        self
    }

    pub fn with_context_metadata(&mut self, meta: ContextMetadata) -> &mut Self {
        self.item.context_metadata = Some(meta);
        self
    }

    pub fn with_client_tag(&mut self, tag: impl Into<String>) -> &mut Self {
        self.item.with_client_tag(tag);
        self
    }

    pub fn build(self) -> ConversationItem {
        self.item
    }
}

pub fn new_tool_call_item(
    id: impl Into<String>,
    name: impl Into<String>,
    args: impl Into<String>,
) -> ToolCallItem {
    ToolCallItem {
        id: id.into(),
        name: name.into(),
        args: args.into(),
        status: ToolCallStatusPending.to_string(),
        description: String::new(),
        streaming_output: String::new(),
        streaming_output_truncated: false,
        result: None,
        error: None,
        duration_ms: 0,
    }
}

pub struct ToolCallItemBuilder {
    item: ToolCallItem,
}

pub fn build_tool_call_item(
    id: impl Into<String>,
    name: impl Into<String>,
    args: impl Into<String>,
) -> ToolCallItemBuilder {
    ToolCallItemBuilder {
        item: new_tool_call_item(id, name, args),
    }
}

impl ToolCallItemBuilder {
    pub fn with_description(&mut self, desc: impl Into<String>) -> &mut Self {
        self.item.description = desc.into();
        self
    }

    pub fn with_status(&mut self, status: impl Into<String>) -> &mut Self {
        self.item.status = status.into();
        self
    }

    pub fn with_streaming_output(
        &mut self,
        output: impl Into<String>,
        truncated: bool,
    ) -> &mut Self {
        self.item.streaming_output = output.into();
        self.item.streaming_output_truncated = truncated;
        self
    }

    pub fn with_result(&mut self, content: impl Into<String>, exit_code: Option<i64>) -> &mut Self {
        self.item.status = ToolCallStatusComplete.to_string();
        self.item.result = Some(ToolCallResult {
            content: content.into(),
            content_truncated: false,
            success: true,
            exit_code,
        });
        self
    }

    pub fn with_error(&mut self, message: impl Into<String>, exit_code: Option<i64>) -> &mut Self {
        self.item.status = ToolCallStatusError.to_string();
        self.item.error = Some(ToolCallError {
            code: String::new(),
            message: message.into(),
            exit_code,
        });
        self
    }

    pub fn with_duration(&mut self, duration_ms: i64) -> &mut Self {
        self.item.duration_ms = duration_ms;
        self
    }

    pub fn build(self) -> ToolCallItem {
        self.item
    }
}

pub fn new_handoff(from_agent: impl Into<String>, to_agent: impl Into<String>) -> ConversationItem {
    ConversationItem {
        item_type: ItemTypeHandoff.to_string(),
        status: ItemStatusComplete.to_string(),
        timestamp: Now(),
        id: String::new(),
        user_input: None,
        turn: None,
        system: None,
        handoff: Some(HandoffInfo {
            from_agent: from_agent.into(),
            to_agent: to_agent.into(),
            tool_name: String::new(),
            input: String::new(),
            reason: String::new(),
        }),
        assistant: None,
        tool_call: None,
        tool_result: None,
        context_metadata: None,
    }
}

pub struct HandoffBuilder {
    item: ConversationItem,
}

pub fn build_handoff(from_agent: impl Into<String>, to_agent: impl Into<String>) -> HandoffBuilder {
    HandoffBuilder {
        item: new_handoff(from_agent, to_agent),
    }
}

impl HandoffBuilder {
    pub fn with_tool_name(&mut self, tool_name: impl Into<String>) -> &mut Self {
        if let Some(handoff) = &mut self.item.handoff {
            handoff.tool_name = tool_name.into();
        }
        self
    }

    pub fn with_input(&mut self, input: impl Into<String>) -> &mut Self {
        if let Some(handoff) = &mut self.item.handoff {
            handoff.input = input.into();
        }
        self
    }

    pub fn with_reason(&mut self, reason: impl Into<String>) -> &mut Self {
        if let Some(handoff) = &mut self.item.handoff {
            handoff.reason = reason.into();
        }
        self
    }

    pub fn with_id(&mut self, id: impl Into<String>) -> &mut Self {
        self.item.id = id.into();
        self
    }

    pub fn build(self) -> ConversationItem {
        self.item
    }
}

pub fn new_system_info(content: impl Into<String>) -> ConversationItem {
    new_system_message(SystemKindInfo.to_string(), content)
}

pub fn new_system_warning(content: impl Into<String>) -> ConversationItem {
    new_system_message(SystemKindWarning.to_string(), content)
}

pub fn new_system_error(content: impl Into<String>) -> ConversationItem {
    new_system_message(SystemKindError.to_string(), content)
}

fn new_system_message(kind: String, content: impl Into<String>) -> ConversationItem {
    ConversationItem {
        item_type: ItemTypeSystem.to_string(),
        status: ItemStatusComplete.to_string(),
        timestamp: Now(),
        id: String::new(),
        user_input: None,
        turn: None,
        system: Some(SystemMessage {
            kind,
            title: String::new(),
            content: content.into(),
        }),
        handoff: None,
        assistant: None,
        tool_call: None,
        tool_result: None,
        context_metadata: None,
    }
}

pub struct SystemBuilder {
    item: ConversationItem,
}

pub fn build_system(kind: impl Into<String>, content: impl Into<String>) -> SystemBuilder {
    let kind = kind.into();
    SystemBuilder {
        item: new_system_message(kind, content),
    }
}

impl SystemBuilder {
    pub fn with_title(&mut self, title: impl Into<String>) -> &mut Self {
        if let Some(system) = &mut self.item.system {
            system.title = title.into();
        }
        self
    }

    pub fn with_id(&mut self, id: impl Into<String>) -> &mut Self {
        self.item.id = id.into();
        self
    }

    pub fn build(self) -> ConversationItem {
        self.item
    }
}

pub fn new_assistant(text: impl Into<String>) -> ConversationItem {
    ConversationItem {
        item_type: ItemTypeAssistant.to_string(),
        status: ItemStatusComplete.to_string(),
        timestamp: Now(),
        id: String::new(),
        user_input: None,
        turn: None,
        system: None,
        handoff: None,
        assistant: Some(Assistant {
            text: text.into(),
            reasoning: String::new(),
            model: String::new(),
            input_tokens: 0,
            output_tokens: 0,
            stop_reason: String::new(),
        }),
        tool_call: None,
        tool_result: None,
        context_metadata: None,
    }
}

pub struct AssistantBuilder {
    item: ConversationItem,
}

pub fn build_assistant(text: impl Into<String>) -> AssistantBuilder {
    AssistantBuilder {
        item: new_assistant(text),
    }
}

impl AssistantBuilder {
    pub fn with_reasoning(&mut self, reasoning: impl Into<String>) -> &mut Self {
        if let Some(assistant) = &mut self.item.assistant {
            assistant.reasoning = reasoning.into();
        }
        self
    }

    pub fn with_model(&mut self, model: impl Into<String>) -> &mut Self {
        if let Some(assistant) = &mut self.item.assistant {
            assistant.model = model.into();
        }
        self
    }

    pub fn with_tokens(&mut self, input: i64, output: i64) -> &mut Self {
        if let Some(assistant) = &mut self.item.assistant {
            assistant.input_tokens = input;
            assistant.output_tokens = output;
        }
        self
    }

    pub fn with_stop_reason(&mut self, reason: impl Into<String>) -> &mut Self {
        if let Some(assistant) = &mut self.item.assistant {
            assistant.stop_reason = reason.into();
        }
        self
    }

    pub fn with_status(&mut self, status: impl Into<String>) -> &mut Self {
        self.item.status = status.into();
        self
    }

    pub fn build(self) -> ConversationItem {
        self.item
    }
}

pub fn new_tool_call(
    call_id: impl Into<String>,
    name: impl Into<String>,
    args: impl Into<String>,
) -> ConversationItem {
    ConversationItem {
        item_type: ItemTypeToolCall.to_string(),
        status: ItemStatusPending.to_string(),
        timestamp: Now(),
        id: String::new(),
        user_input: None,
        turn: None,
        system: None,
        handoff: None,
        assistant: None,
        tool_call: Some(ToolCall {
            call_id: call_id.into(),
            name: name.into(),
            args: args.into(),
            description: String::new(),
        }),
        tool_result: None,
        context_metadata: None,
    }
}

pub struct ToolCallBuilder {
    item: ConversationItem,
}

pub fn build_tool_call(
    call_id: impl Into<String>,
    name: impl Into<String>,
    args: impl Into<String>,
) -> ToolCallBuilder {
    ToolCallBuilder {
        item: new_tool_call(call_id, name, args),
    }
}

impl ToolCallBuilder {
    pub fn with_description(&mut self, desc: impl Into<String>) -> &mut Self {
        if let Some(tool_call) = &mut self.item.tool_call {
            tool_call.description = desc.into();
        }
        self
    }

    pub fn with_status(&mut self, status: impl Into<String>) -> &mut Self {
        self.item.status = status.into();
        self
    }

    pub fn build(self) -> ConversationItem {
        self.item
    }
}

pub fn new_tool_result(
    call_id: impl Into<String>,
    content: impl Into<String>,
    is_error: bool,
) -> ConversationItem {
    ConversationItem {
        item_type: ItemTypeToolResult.to_string(),
        status: ItemStatusComplete.to_string(),
        timestamp: Now(),
        id: String::new(),
        user_input: None,
        turn: None,
        system: None,
        handoff: None,
        assistant: None,
        tool_call: None,
        tool_result: Some(ToolResult {
            call_id: call_id.into(),
            content: content.into(),
            is_error,
            exit_code: None,
            streaming_output: String::new(),
            output_truncated: false,
            duration_ms: 0,
        }),
        context_metadata: None,
    }
}

pub struct ToolResultBuilder {
    item: ConversationItem,
}

pub fn build_tool_result(
    call_id: impl Into<String>,
    content: impl Into<String>,
) -> ToolResultBuilder {
    ToolResultBuilder {
        item: new_tool_result(call_id, content, false),
    }
}

impl ToolResultBuilder {
    pub fn with_error(&mut self) -> &mut Self {
        if let Some(result) = &mut self.item.tool_result {
            result.is_error = true;
        }
        self
    }

    pub fn with_exit_code(&mut self, code: i64) -> &mut Self {
        if let Some(result) = &mut self.item.tool_result {
            result.exit_code = Some(code);
        }
        self
    }

    pub fn with_streaming_output(&mut self, output: impl Into<String>) -> &mut Self {
        if let Some(result) = &mut self.item.tool_result {
            result.streaming_output = output.into();
        }
        self
    }

    pub fn with_truncated(&mut self) -> &mut Self {
        if let Some(result) = &mut self.item.tool_result {
            result.output_truncated = true;
        }
        self
    }

    pub fn with_duration(&mut self, duration_ms: i64) -> &mut Self {
        if let Some(result) = &mut self.item.tool_result {
            result.duration_ms = duration_ms;
        }
        self
    }

    pub fn build(self) -> ConversationItem {
        self.item
    }
}

pub fn attach_provenance(meta: &mut ContextMetadata, provenance: Provenance) {
    meta.provenance = Some(provenance);
}
